rclcpp/include/rclcpp/network_flow_endpoint.hpp

```diff
@@ -0,0 +1,108 @@
+// Copyright 2020 Ericsson AB
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__NETWORK_FLOW_ENDPOINT_HPP_
+#define RCLCPP__NETWORK_FLOW_ENDPOINT_HPP_
+
+#include <cstdint>
+#include <string>
+#include <iostream>
+
+#include "rcl/network_flow_endpoints.h"
+
+#include "rclcpp/visibility_control.hpp"
+
+namespace rclcpp
+{
+
+/**
+ * Class describes a network flow endpoint based on the counterpart definition
+ * in the RMW layer.
+ */
+class NetworkFlowEndpoint
+{
+public:
+  /// Construct from rcl_network_flow_endpoint_t
+  RCLCPP_PUBLIC
+  explicit NetworkFlowEndpoint(const rcl_network_flow_endpoint_t & network_flow_endpoint)
+  : transport_protocol_(
+      rcl_network_flow_endpoint_get_transport_protocol_string(network_flow_endpoint.
+      transport_protocol)),
+    internet_protocol_(
+      rcl_network_flow_endpoint_get_internet_protocol_string(
+        network_flow_endpoint.internet_protocol)),
+    transport_port_(network_flow_endpoint.transport_port),
+    flow_label_(network_flow_endpoint.flow_label),
+    dscp_(network_flow_endpoint.dscp),
+    internet_address_(network_flow_endpoint.internet_address)
+  {
+  }
+
+  /// Get transport protocol
+  RCLCPP_PUBLIC
+  const std::string & transport_protocol() const;
+
+  /// Get internet protocol
+  RCLCPP_PUBLIC
+  const std::string & internet_protocol() const;
```

Comment:
I wonder, should these be stored as fixed sized c-strings (or maybe a string_view to a c-string) and returned as `const char *` (or again as string_view)?

They never change and their maximum size is already known, so it seems like `std::string` is overkill?

Reply:
Thought about this and tried a couple of approaches -- guarding the member string as const and using a char buffer. Both made me uneasy. The constant member string upset move semantics. The char buffer became unwieldy due to unclear max size and increased coupling with rmw. 

I prefer to keep it as std::string and relay on optimizations (e.g. SBO) by the compiler to retain speed. The interface is also not on the typical critical path and is most likely going to be used once during the lifetime of a publisher/subscription. 

When C++17 becomes the norm for ROS2 packages, the std::string container can be switched to std::string_view.

Reply:
Ok, technically that's already the case (we're using `std::variant` now), but we can wait.